blog/_posts/2021-05-18-argocd-private-clusters.md

```diff
@@ -0,0 +1,414 @@
+namespace=$(kubectl get --context $TARGET_CTX -n kube-system secret/$name -o jsonpath='{.data.namespace}' | base64 --decode)
+
+cat <<EOF | kubectl apply --context $ARGOCD_CTX -n argocd -f -
+apiVersion: v1
+kind: Secret
+metadata:
+  name: gke-eu1
+  labels:
+    argocd.argoproj.io/secret-type: cluster
+type: Opaque
+stringData:
+  name: gke-eu1
+  server: https://gke-eu1.inlets:443
+  config: |
+    {
+      "bearerToken": "${token}",
+      "tlsClientConfig": {
+        "insecure": true
+      }
+    }
+EOF
+```
+
+Notice the `"insecure" : true"`? This is probably necessary as the certificate of the Kubernetes API service is not valid for the domain name `gke-eu1.inlets`. Where we have access to kubeadm or k3s for the private cluster, we can add a TLS SAN name, and the solution works by directly tunnelling the API server.
+
+![argocd](/images/2021-05-18-argocd-private-clusters/argocd_clusters.png)
+> Our private cluster is ready to serve!
+
+## Giving it a test
+
+Now it is time to deploy an application.
+
+Let's try to deploy the guestbook application in a declarative way:
+
+``` bash
+export ARGOCD_CTX=argocd
+
+cat <<EOF | kubectl apply --context $ARGOCD_CTX -n argocd -f -
+apiVersion: argoproj.io/v1alpha1
+kind: Application
+metadata:
+  name: guestbook
+  namespace: argocd
+spec:
+  destination:
+    namespace: default
+    server: https://gke-eu1.inlets:443
+  project: default
+  source:
+    path: kustomize-guestbook
+    repoURL: https://github.com/argoproj/argocd-example-apps.git
+    targetRevision: HEAD
+  syncPolicy:
+    automated: {}
+EOF
+```
+
+Or if you have the ArgoCD cli available, why won't you try installing OpenFaas with ArgoCD?
+
+``` bash
+export TARGET_CTX=gke-eu1
+
+kubectl create --context $TARGET_CTX namespace openfaas
+kubectl create --context $TARGET_CTX namespace openfaas-fn
+
+argocd app create openfaas-operator \
+      --repo https://github.com/openfaas/faas-netes.git \
+      --dest-namespace openfaas \
+      --dest-server https://gke-eu1.inlets:443 \
+      --path chart/openfaas \
+      --helm-set operator.create=true \
+      --helm-set generateBasicAuth=true \
+      --helm-set functionNamespace=openfaas-fn \
+      --self-heal \
+      --sync-policy automatic
+```
+
+As soon those commands are applied, you will notice Argo CD starts synchronisation the desired state to the target cluster.
+And a few moments later, you're applications are up and running!
+
+![argocd](/images/2021-05-18-argocd-private-clusters/argocd_application.png)
+> Argo CD installing the guestbook and OpenFaas in the private GKE cluster
+
+## Closing thoughts
+
+And that's a wrap! 
+
+You should now have a clear idea of how inlets PRO can help to bring sensitive services, such as a private Kubernetes API service, into a central management cluster in a secure way. Using this technique, the control plane of the tunnel is accessible by the clients, while keeping the data plane secure and private in the cluster. 
```

Comment:
Then, reiterate the problem and then go on to mention how it was solved
* multiple public / private clusters
* potentially in different cloud vendors
* need to deploy centrally for ease of use and monitoring the state of applications
* modern GitOps approach is still possible 
* 
* "You should now have a clear idea" <- it may be clear to you and me, others may not agree.

It can be written differently describing what took place, rather than how much the user understood. Just a nuance in the meaning.